In [1]:
import pandas as pd
from neuronunit.optimization.optimization_management import public_opt
from sciunit.scores import RelativeDifferenceScore, ZScore
from sciunit.utils import config_set, config_get
config_set('PREVALIDATE', False)
from neuronunit.allenapi.allen_data_driven import make_allen_hard_coded_limited, make_allen_hard_coded_complete
import warnings
from neuronunit.optimization.model_parameters import BPO_PARAMS
warnings.filterwarnings("ignore")
_,_,_,a_cells = make_allen_hard_coded_limited()


[
    allen_suite_623960880,
    allen_suite_623893177,
    allen_suite471819401,
    allen_suite482493761,
    a_cells2,
    df
] = make_allen_hard_coded_complete()



/home/user/anaconda3/lib/python3.7/site-packages/cerberus/validator.py:1609: UserWarning: No validation schema is defined for the arguments of rule 'not_zero_obs_zscore'
  "'%s'" % method_name.split('_', 2)[-1]
test


# In this notebook
we fit Izhikevich and Adaptive Exponential Integrate and fire models to electrophysiology observations 
from the Allen Brain, using BluePyOpt and NeuronUnit.

* Immediately Below 
is a list of some of the observations we will use:

In [2]:
df

TimeConstantTest       name RestingPotentialTest  \
623960880          23.8 ms  623960880             -65.1 mV   
623893177          27.8 ms  623893177             -77.0 mV   
471819401          13.8 ms  471819401             -77.5 mV   
482493761          24.4 ms  482493761             -71.6 mV   

          InputResistanceTest RheobaseTest  
623960880       241.0 megaohm      70.0 pA  
623893177       136.0 megaohm     190.0 pA  
471819401       132.0 megaohm     190.0 pA  
482493761       132.0 megaohm      70.0 pA

Allen Specimen ID: 471819401
Four data fitting constraints:
* RheobaseTest	
* TimeConstantTest	
* RestingPotentialTest	
* InputResistanceTest	
# Model AdExp

In [3]:
opt,obs_preds,df = public_opt(
    a_cells['471819401'],
    BPO_PARAMS,
    '471819401',
    "ADEXP",
    100,
    200,
    "IBEA",
    score_type=RelativeDifferenceScore
    )


gen	nevals	avg    	std    	min    	max 
1  	100   	1538.18	1514.29	12.4581	4000


2  	45    	1495.88	1478.09	12.4581	4000
3  	8     	521.317	693.355	12.4581	4000
4  	9     	169.522	361.724	11.7822	1010.05
5  	7     	345.401	468.199	11.7822	1009.93
6  	6     	558.644	494.769	11.7822	1007.78
7  	9     	597.779	489.112	9.83394	1007.67
8  	7     	421.771	488.943	9.83394	1006.98
9  	11    	343.136	469.018	9.74488	1008.07
10 	14    	421.319	488.847	9.74488	1007.56
11 	11    	440.64 	491.915	8.8705 	1008.69
12 	5     	264.488	432.584	8.8705 	1006.71
13 	10    	536.616	495.672	5.4392 	1007.94
14 	8     	438.925	492.656	5.4392 	1007.94
15 	6     	185.129	379.489	5.4392 	1006.17
16 	10    	244.933	421.679	5.4392 	1007.21
17 	7     	418.84 	490.367	5.4392 	1005.93
18 	12    	419.517	490.006	5.4392 	1007.43
19 	12    	536.703	497.164	5.4392 	1008.4 
20 	12    	341.273	469.346	5.4392 	1007.4 
21 	6     	183.519	380.349	5.4392 	1006.2 
22 	12    	380.979	481.393	5.4392 	1008.41
23 	15    	360.138	476.521	5.4392 	1007.95
24 	12    	223.606	409.826	5.4392 	1006.6 
25 	17    	457.52

193	13    	298.688	454.391	2.2118 	1003.54
194	6     	435.662	494.107	2.2118 	1004.74
195	8     	239.215	423.321	2.2118 	1003.54
196	14    	533.509	497.649	2.2118 	1004.83
197	16    	258.615	433.973	2.2118 	1001.65
198	8     	181.568	380.063	2.2118 	1005.72
199	10    	62.9985	234.404	2.2118 	1001.36
200	7     	121.743	321.209	2.2118 	1007.05
Run stopped because of stopping criteria: Max ngen


In [4]:
df

observations    predictions
RheobaseTest               190.0 pA      158.61 pA
TimeConstantTest            13.8 ms       11.05 ms
RestingPotentialTest       -77.5 mV      -85.53 mV
InputResistanceTest   132.0 megaohm  98.83 megaohm

# Solution Model Parameters:


In [5]:
opt.attrs

{'cm': 137.67833344035898,
 'v_spike': -35.13275001670514,
 'v_reset': -40.74087751009728,
 'v_rest': -77.60308737269521,
 'tau_m': 47.679090765165924,
 'a': 5.699447034775152,
 'b': 2.5472852479489463,
 'delta_T': 9.615925026124955,
 'tau_w': 163.56605801714784,
 'v_thresh': -19.32533012991194,
 'spike_delta': 91.39955649682773}

Allen Specimen ID: 471819401
Four data fitting constraints:
* RheobaseTest	
* TimeConstantTest	
* RestingPotentialTest	
* InputResistanceTest	
# Model Izhikevich

In [6]:
opt,obs_preds,df = public_opt(
    a_cells['471819401'],
    BPO_PARAMS,
    'fi_curve',
    "IZHI",
    100,
    360,
    "IBEA",
    )

gen	nevals	avg   	std    	min    	max 
1  	100   	1408.1	1110.01	13.2802	4000


2  	53    	1346.63	1100.44	11.775 	4000
3  	8     	735.313	443.628	11.775 	1009.73
4  	9     	676.292	468.296	11.775 	1009.6 
5  	10    	636.806	480.257	11.775 	1009.02
6  	13    	617.109	485.907	10.7102	1009.7 
7  	9     	675.908	468.342	10.7102	1009.75
8  	4     	909.191	295.311	10.7102	1008.83
9  	14    	715.054	452.765	9.91486	1009.34
10 	12    	928.817	267.028	9.91486	1009.06
11 	14    	792.902	408.624	9.91486	1009.56
12 	8     	713.578	452.934	9.77148	1008.92
13 	8     	695.458	460.424	9.77148	1008.34
14 	8     	752.959	433.972	9.77148	1008.11
15 	14    	791.466	409.587	9.77148	1008.12
16 	9     	869.489	342.506	9.77148	1008.02
17 	14    	732.583	443.802	9.77148	1008.81
18 	11    	927.043	267.225	9.77148	1007.31
19 	8     	790.827	409.104	9.77148	1007.77
20 	14    	829.697	379.028	9.77148	1007.98
21 	5     	771.285	421.713	9.70916	1006.61
22 	14    	927.272	267.325	9.70916	1007.17
23 	13    	731.609	443.416	9.70916	1008.34
24 	9     	810.303	395.145	9.70916	1007.01
25 	8     	712

193	9     	905.449	294.856	8.59166	1006.16
194	10    	730.021	442.765	8.59166	1003.99
195	6     	186.181	378.07 	8.59166	1008.48
196	14    	478.548	494.669	8.59166	1008.25
197	16    	594.943	488.076	8.59166	1008.07
198	14    	827.936	377.734	8.59166	1008.27
199	7     	555.385	493.743	8.59166	1008.48
200	6     	89.4735	267.459	8.59166	1008.3 
201	8     	265.378	433.937	8.59166	1008.61
202	8     	557.9  	495.653	8.59166	1008.51
203	11    	518.422	496.463	8.59166	1008.29
204	14    	711.322	451.672	8.59166	1008.06
205	12    	380.64 	478.83 	8.59166	1008.06
206	11    	652.579	474.234	8.11064	1008.11
207	8     	224.904	408.472	8.11064	1008.05
208	8     	594.656	488.866	8.11064	1008.06
209	10    	593.128	489.464	8.11064	1007.4 
210	10    	904.592	295.564	8.11064	1007.48
211	9     	729.101	443.5  	8.11064	1007.15
212	7     	300.676	453.241	8.11064	1007.69
213	12    	709.769	452.921	8.11064	1007.47
214	11    	242.529	422.377	8.11064	1008.11
215	12    	594.601	490.097	8.11064	1007.69
216	9     	

In [7]:
df

observations     predictions
RheobaseTest               190.0 pA       420.69 pA
TimeConstantTest            13.8 ms        14.92 ms
RestingPotentialTest       -77.5 mV       -74.76 mV
InputResistanceTest   132.0 megaohm  127.42 megaohm

# Solution Parameters:

In [8]:
opt.attrs

{'C': 134.8647638028551,
 'k': 0.2519686430005483,
 'vr': -70.68407968876562,
 'vt': -42.32491031269837,
 'vPeak': 0.9681701933782172,
 'a': 0.035121443074930404,
 'b': 0.3997394531945915,
 'c': -54.89915700512454,
 'd': 89.47106917587624,
 'celltype': 4.136832661561244}

# Just FI curve on AdExp model
Allen Specimin Id: 471819401

In [9]:
opt,obs_preds,df = public_opt(
    a_cells['fi_curve'],
    BPO_PARAMS,
    'fi_curve',
    "ADEXP",
    100,
    100,
    "IBEA"
    )

gen	nevals	avg    	std    	min    	max 
1  	100   	1651.54	495.397	4.32057	2000


2  	45    	1633.36	538.632	3.63816	2000
3  	8     	1139.22	685.128	2.63703	2000
4  	9     	592.627	631.342	2.63703	2000
5  	7     	238.927	579.328	2.63703	2000
6  	6     	61.9963	306.972	2.63703	2000
7  	9     	22.1949	138.295	2.63703	1000.09
8  	9     	22.1949	138.295	2.63703	1000.09
9  	6     	2.63703	4.44089e-16	2.63703	2.63703
10 	9     	61.3092	234.689    	2.63703	1000.09
11 	12    	22.1949	138.295    	2.63703	1000.09
12 	8     	2.63703	4.44089e-16	2.63703	2.63703
13 	6     	2.63703	4.44089e-16	2.63703	2.63703
14 	15    	41.801 	276.931    	2.63703	2000   
15 	16    	22.1934	138.284    	2.63703	1000.01
16 	8     	2.63703	4.44089e-16	2.63703	2.63703
17 	8     	2.63703	4.44089e-16	2.63703	2.63703
18 	13    	41.801 	276.931    	2.63703	2000   
19 	6     	2.63703	4.44089e-16	2.63703	2.63703
20 	15    	22.3025	138.896    	2.63703	1004.44
21 	15    	22.1936	138.286    	2.63703	1000.02
22 	8     	2.63703	4.44089e-16	2.63703	2.63703
23 	10    	2.63703	4.44089e-16	2.63703	2.63703
24 	4    

In [10]:
df


observations  predictions
RheobaseTest     190.0 pA    189.21 pA
FITest         0.09 Hz/pA  [0.0 Hz/pA]

In [11]:
opt.attrs

{'cm': 527.8153065477139,
 'v_spike': -46.28357188950913,
 'v_reset': -76.10986130705929,
 'v_rest': -43.344602017468915,
 'tau_m': 53.57336806011064,
 'a': 1.7802197513620552,
 'b': 2.704097569227171,
 'delta_T': 3.9385659154088506,
 'tau_w': 338.10347759692854,
 'v_thresh': -21.279394701940934,
 'spike_delta': 63.6931103717196}

Allen Specimen ID: 623960880
Four data fitting constraints:
* RheobaseTest	
* TimeConstantTest	
* RestingPotentialTest	
* InputResistanceTest	
# Model Izhikevich

In [12]:

opt,obs_preds,df = public_opt(
    allen_suite_623960880.tests,
    BPO_PARAMS,
    allen_suite_623960880.name,
    "IZHI",
    100,
    100,
    "IBEA"
    )

gen	nevals	avg    	std    	min    	max 
1  	100   	1408.55	1109.79	13.5395	4000


2  	53    	1347.1 	1100.21	13.5395	4000
3  	8     	697.455	460.935	13.5395	1016.53
4  	10    	715.706	452.873	13.4101	1016.53
5  	14    	910.388	295.403	11.9954	1016.13
6  	12    	559.879	494.759	11.9954	1010.39
7  	10    	715.321	452.979	11.9954	1009.75
8  	12    	675.753	469.105	11.9954	1010.01
9  	11    	948.888	234.019	11.9954	1010.41
10 	8     	851.151	361.497	11.9954	1010.38
11 	12    	733.858	443.726	11.9954	1009.9 
12 	10    	636.544	480.863	11.9954	1009.98
13 	14    	831.612	379.071	11.9954	1010.52
14 	7     	675.614	468.869	11.9954	1008.06
15 	7     	987.383	137.942	11.9954	1007.62
16 	8     	968.271	193.198	11.9954	1010.45
17 	7     	968.099	193.163	11.9954	1008.72
18 	13    	870.233	342.341	11.361 	1009.55
19 	11    	791.867	408.954	11.361 	1007.22
20 	8     	752.943	433.571	11.361 	1007.22
21 	9     	830.582	379.227	11.361 	1007.11
22 	10    	791.986	409.363	11.361 	1007.12
23 	13    	792.495	409.631	11.361 	1010.44
24 	20    	870.594	342.633	11.361 	1010.24
25 	12    	792

In [13]:
df

observations   predictions
TimeConstantTest            23.8 ms       4.21 ms
RestingPotentialTest       -65.1 mV     -65.16 mV
InputResistanceTest   241.0 megaohm  58.4 megaohm
RheobaseTest                70.0 pA      73.56 pA

In [14]:
opt.attrs

{'C': 23.348910505873352,
 'k': 0.8098168647307745,
 'vr': -60.62852680673786,
 'vt': -49.68123470796534,
 'vPeak': 17.874565331816484,
 'a': 0.10812123709137333,
 'b': 12.858661006399277,
 'c': -59.092708738853666,
 'd': 31.844623471110687,
 'celltype': 3.4141675865786247}